refactor: extract and move comment formatting logic to util

emmileaf opened the pull request:
Both `ClientLibraryPackageInfoComposer` and `ServiceClientCommentComposer` currently have additional parsing logic to convert protobuf comments to a Javadoc-compatible format. For example, this handles [bulleted list items with asterisks in secret manager service proto](https://github.com/googleapis/googleapis/blob/db1fc8a59aeec7aabfb3dfed4ce9944a312ed55e/google/cloud/secretmanager/v1/service.proto#L44-L45). 

This PR extracts common logic to an util class, and exposes it for reuse by Spring Codegen.

diegomarquezp reviewed (approved):
LGTM - nice cleanup!

blakeli0 commented on src/main/java/com/google/api/generator/gapic/composer/utils/CommentFormatter.java:
```diff
@@ -0,0 +1,62 @@
+
+public class CommentFormatter {
+
+  private CommentFormatter() {}
```
This empty constructor is not needed.

zhumin8 replied:
I think this is from sonar recommendation to use private constructor to hide the implicit public one.
https://rules.sonarsource.com/java/RSPEC-1118

blakeli0 replied:
I see, this is one of the controversial rules though, see this [stackoverflow](https://stackoverflow.com/questions/14398747/how-to-disable-warning-on-sonar-hide-utility-class-constructor) discussion. IMO, it does not provide much value and add unnecessary code, maybe we should turn off this rule? WDYT @meltsufin ?

meltsufin replied:
I kind of agree with Sonar here. A private constructor is helpful to avoid mistakes around instantiating the class when it's intended to be used as a collection of static utility methods.

blakeli0 replied:
I think it makes sense if this is a util class in `gax-java` or a backend of a web app, but maybe not for `gapic-generator-java`, because pretty much everything in the generator is called through `public static` methods in parsers and composers, the whole generator is basically a util. 

There may be unnecessary garbage collection if we accidentally instantiated a util class, which could cause some harm in a server if the util method is call multiple times, but there is no harm for the generator because the generator is run once per jvm.

blakeli0 replied:
Mike and I discussed offline and we are going to remove this rule from gapic-generator-java, but it's not critically enough to block this PR.

blakeli0 commented on src/main/java/com/google/api/generator/gapic/composer/utils/CommentFormatter.java:
```diff
@@ -0,0 +1,62 @@
+  private CommentFormatter() {}
+
+  // Additional formatting logic for converting protobuf comment to Javadoc
+  public static String formatAndAddToJavaDocComment(String comment, String prefixPattern) {
```
The method name may not reflect the latest changes, usually if you have an `and` in a method name, it's a code smell that either the method is doing too much or you can have a better name, maybe something like `fromProtobufCommentToJavaDocComment`?(I'm not good at naming as well, just throwing out an idea here)

emmileaf replied:
Thanks for catching this, I missed updating the method name when switching to returning a string instead of adding to a builder. Renamed to `formatAsJavaDocComment`.

blakeli0 commented on src/main/java/com/google/api/generator/gapic/composer/utils/CommentFormatter.java:
```diff
@@ -0,0 +1,62 @@
+
+  private CommentFormatter() {}
+
+  // Additional formatting logic for converting protobuf comment to Javadoc
```
Can you provide a little more example about what exact this method is doing?

emmileaf replied:
Updated the unit tests to a couple of different parsing cases - hopefully this illustrates the method's behaviour a bit better!

blakeli0 reviewed (changes requested):
Please resolve the comments

emmileaf commented:
Regarding the code coverage:

The use of this logic/util method in the two composers is covered by the integration tests (e.g. [redis golden file](https://github.com/googleapis/gapic-generator-java/blob/main/test/integration/goldens/redis/src/com/google/cloud/redis/v1beta1/package-info.java)). Sonar is unhappy with coverage because the existing composer unit tests (e.g. through `ServiceClientClassComposerTest`) do not hit any of this logic -  it only applies when services or methods have a `.description()` field, which showcase services do not.
 
Todo: will need to look more into how `Parser` gets descriptions for services and methods to see if unit test resources can be updated to cover comment parsing in existing Composer tests, or if this behavior needs to be separately mocked.
- [service.description()](https://github.com/googleapis/gapic-generator-java/blob/ee5d7a5985d298e6d96dc700805870f3e362b1de/src/main/java/com/google/api/generator/gapic/protoparser/Parser.java#L457-L464)
- [method.description()](https://github.com/googleapis/gapic-generator-java/blob/ee5d7a5985d298e6d96dc700805870f3e362b1de/src/main/java/com/google/api/generator/gapic/protoparser/Parser.java#L686-L693)

emmileaf commented:
Some additional test-related changes from addressing code coverage:
- Added description for echo service explicitly in `TestProtoLoader` since it is not parsed from source code location as an end-to-end call would
  - This allows the existing `ServiceClientClassTest` to cover class header comments that are also formatted with the util’s logic
- Added `ClientLibraryPackageInfoComposerTest` as a golden-updating test
